### PR TITLE
Define `GPU_COMPUTE_CAPABILITY`

### LINKED ISSUE #207: Define GPU_COMPUTE_CAPABILITY
Define `GPU_COMPUTE_CAPABILITY` to replace `GPU_ARCH` in the `Makefile`. For example, set `GPU_COMPUTE_CAPABILITY=860` to replace `GPU_ARCH=AMPERE`. It gives more precise control of the adopted [GPU compute capability](https://developer.nvidia.com/cuda-gpus) since, for example, an Ampere GPU can map to more than one compute capability (`800, 860, 870`).

### PR DESCRIPTION
**Pull request by ChunYen-Chen**
Resolves #207.

### New feature
An auto-detect GPU compute capability function is given in `configure.py`.

### CODE REVIEW

**Review by hyschive — CHANGES_REQUESTED** [review]
@ChunYen-Chen Good job. In addition to the inline comments, I suggest the following changes/tasks.
- In `CUAPI_SetDevice.cu`, revise the sanity checks on `GPU_ARCH` to compare `GPU_COMPUTE_CAPABILITY` and `DeviceProp.major/minor` directly ([example](https://github.com/hyschive/gamer-fork/blob/psidm/src/GPU_API/CUAPI_SetDevice.cu#L204-L208))
- Update these [comments](https://github.com/gamer-project/gamer/blob/main/example/queue/submit_ncts_intel.job#L45-L51) about mixing TURING and AMPERE on the NCTS cluster (@hfhsieh can help validate it)
- Record `GPU_COMPUTE_CAPABILITY` in `Aux_TakeNote.cpp`
- Once this PR is merged, file a follow-up PR to merge `main` into `psidm`
  - Remove `GPU_COMPUTE_CAPABILITY` from `CUFLU.h` 
  - Update or remove the error message “Please update GPU_COMPUTE_CAPABILITY in CUFLU.h” in `CUAPI_SetDevice.cu`

**Comment by hfhsieh** [comment]
@hyschive I have not updated this file yet. However, these [comments](https://github.com/gamer-project/gamer/blob/main/example/queue/submit_ncts_intel.job#L45-L51) are deprecated since the computer architecture on the NCTS cluster have been updated (collect the same GPUs into the same node).

**Comment by hyschive** [comment]
> @hyschive I have not updated this file yet. However, these [comments](https://github.com/gamer-project/gamer/blob/main/example/queue/submit_ncts_intel.job#L45-L51) are deprecated since the computer architecture on the NCTS cluster have been updated (collect the same GPUs into the same node).

@hfhsieh Thanks for the clarification.
@ChunYen-Chen Then you can simply remove those comments.

File: src/configure.py [hunk 1]
```diff
@@ -275,6 +276,59 @@ def distance( s1, s2 ):
 
     return matrix[len(s1)][len(s2)]
 
+def get_gpu_compute_capability():
```
[the remark this answers is not in the document]
Comment (ChunYen-Chen) [reply src/configure.py:277]: The error is raised correctly for the machine without CUDA. I do not have a device without an NVIDIA GPU, so I did not test it.

**Comment by ChunYen-Chen** [comment]
@hyschive Thanks for your review! I have modified the script according to the comments. I do not have the machine without any NVIDIA GPU, so I leave the conversion about `get_gpu_compute_capability()` unresolved.

File: src/configure.py [hunk 2]
```diff
@@ -749,15 +744,45 @@ def set_conditional_defaults( args ):
...
+    return gpu_opts
+
 def set_sims( name_table, depends, **kwargs ):
     opt_str = ""
-#   Loop all the simulation options in GAMER.
+    # loop all the simulation options in GAMER.
     for opt, gamer_name in name_table.items():
         store = True
-#       check if the depend is true
+        # check if the depend is true
```
Comment (hyschive) [inline src/configure.py:785]: Remove "the"

File: src/configure.py [hunk 3]
```diff
@@ -769,21 +794,21 @@ def set_sims( name_table, depends, **kwargs ):
...
+    # hard-code the option of serial.
     if not kwargs["mpi"]: opt_str = add_option( opt_str, name="SERIAL", val=True )
 
     return {"SIMU_OPTION":opt_str}
 
 def set_compile( paths, compilers, flags, kwargs ):
     com_opt = {}
 
-#   1. Set the complier.
+    # 1. Set the complier.
```
Comment (hyschive) [inline src/configure.py:805]: Typo: complier

File: src/configure.py [hunk 4]
```diff
@@ -769,21 +794,21 @@ def set_sims( name_table, depends, **kwargs ):
...
     com_opt["CXX"] = os.path.join(paths["MPI_PATH"], "bin", compilers["CXX_MPI"]) if kwargs["mpi"] else compilers["CXX"]
 
-#   2. Set the OpenMP flags.
+    # 2. Set the OpenMP flags.
     if not kwargs["openmp"]: flags["OPENMPFLAG"] = ""
 
-#   3. Write flags to complie option dictionary.
+    # 3. Write flags to complie option dictionary.
```
Comment (hyschive) [inline src/configure.py:811]: Typo: complie

File: src/configure.py [hunk 5]
```diff
@@ -901,42 +925,45 @@ def warning( paths, **kwargs ):
 # Main execution
 ####################################################################################################
 if __name__ == "__main__":
-#   0. Setup logger
+    # 0. Setup logger
```
Comment (hyschive) [inline src/configure.py:928]: "Set the logger"

File: src/configure.py [hunk 6]
```diff
@@ -901,42 +925,45 @@ def warning( paths, **kwargs ):
...
     ch = logging.StreamHandler()
     ch.setFormatter( CustomFormatter() )
     LOGGER.addHandler( ch )
 
-#   1. Get the executed command
+    # 1. Get the executed command
```
Comment (hyschive) [inline src/configure.py:933]: "executed" --> "execution"

File: src/configure.py [hunk 7]
```diff
@@ -901,42 +925,45 @@ def warning( paths, **kwargs ):
...
 
-#   3.3 add the SIMU_OPTION
+    # 3.3 Add the SIMU_OPTION
     LOGGER.info("========================================")
     LOGGER.info("GAMER has the following setting.")
     LOGGER.info("----------------------------------------")
     sims = set_sims( name_table, depends, **args )
 
-#   3.4 setup compiler
+    # 3.4 Setup compiler
```
Comment (hyschive) [inline src/configure.py:955]: "Set the compiler"

File: src/configure.py [hunk 8]
```diff
@@ -901,42 +925,45 @@ def warning( paths, **kwargs ):
...
     compiles = set_compile( paths, compilers, flags, args )
 
-#   4. Create Makefile
-#   4.1 Read
+    # 3.5 Setup gpus
```
Comment (hyschive) [inline src/configure.py:958]: "Set the GPU"

File: src/configure.py [hunk 9]
```diff
@@ -792,12 +817,12 @@ def set_compile( paths, compilers, flags, kwargs ):
 def validation( paths, depends, constraints, **kwargs ):
     success = True
 
-#   0. Checking the Makefile_base existance.
+    # 0. Checking the Makefile_base existance.
```
Comment (hyschive) [inline src/configure.py:820]: typo: existance

**Review by hyschive — CHANGES_REQUESTED** [review]
- Set the following Compute Capabilities.
  - `cannon.config`:
    ```
    GPU_COMPUTE_CAPABILITY 800    # NVIDIA A100 
    #GPU_COMPUTE_CAPABILITY 700    # NVIDIA V100
    ```
  - `fomalhaut.config`:
    ```
    GPU_COMPUTE_CAPABILITY 860    # GeForce RTX 3080 and RTX A4000
    #GPU_COMPUTE_CAPABILITY 750    # GeForce RTX 2080 Ti 
       ```
  - `ncts.config`:
    ```
    GPU_COMPUTE_CAPABILITY 860    # RTX A6000 
    #GPU_COMPUTE_CAPABILITY 750    # Quadro RTX 6000
    ```
  - `pleiades.config`:
    ```
    GPU_COMPUTE_CAPABILITY 700    # NVIDIA V100
    ```
- Remove `bluewaters.config` as this system has been decommissioned
- For consistency, how about setting `NVCCFLAG_COM, NVCCFLAG_FLU, NVCCFLAG_POT` completely in `configure.py`, except for `NVCCFLAG_COM += $(COMMONFLAG)`?
- There are many typos in `configure.py`. Check it carefully.
- When enabling `--gpu`, raise an error instead of a warning if `GPU_COMPUTE_CAPABILITY` is not set properly (e.g., when it's not set in `*.config`); you can refer to `../configs/template.config` in the error message.
- Align the following lines in `Makefile.log`
```
GPU_COMPUTE_CAPABILITY : 860
NVCCFLAG_ARCH   : -gencode arch=compute_86,code=\"compute_86,sm_86\"
NVCCFLAG_FLU    : --maxrregcount=192
```

File: src/configure.py [hunk 10]
```diff
@@ -746,43 +746,45 @@ def set_conditional_defaults( args ):
...
 
-def set_gpu( gpus, args ):
+def set_gpu( gpus, flags, args ):
     gpu_opts = {}
     compute_capability = gpus["GPU_COMPUTE_CAPABILITY"]
 
     # 1. Check the compute capability
-    if compute_capability == "": return gpu_opts
+    if compute_capability == "":
+        if args["gpu"]: raise ValueError("GPU_COMPUTE_CAPABILITY is not set in *.config. Please checkout `../configs/template.config` to see how to set GPU_COMPUTE_CAPABILITY.")
```
Comment (hyschive) [inline src/configure.py:755]: How about changing the error message to
```
"GPU_COMPUTE_CAPABILITY is not set in `../configs/%s.config`. See `../configs/template.config` for illustration." % args["machine"]
```

File: src/configure.py [hunk 11]
```diff
@@ -315,8 +315,7 @@ def cuda_check_error( result ):
     cuda_check_error( cuda.cuInit(0) )
     cuda_check_error( cuda.cuDeviceGetCount(ctypes.byref(nGpus)) )
 
-    arch = ""
-    if nGpus.value > 1: print("WARNING: More than one GPU. Selecting the last GPU architecture.")
+    if nGpus.value > 1: print("WARNING: More than one GPU. Selecting the last GPU compute capability.")
```
Comment (hyschive) [inline src/configure.py:318]: Change the warning message to "WARNING: More than one GPU --> select the compute capability of the last GPU."

File: src/configure.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hyschive) [reply src/configure.py:277]: @ChunYen-Chen Have you tested it on a system without any NVIDIA GPU (e.g., on a MacBook)?

**Review by hyschive — COMMENTED** [review]
- `GPU_COMPUTE_CAPABILITY` is still missing in several `*.config`. Check my comment [here](https://github.com/gamer-project/gamer/pull/302#pullrequestreview-2063439922) carefully.
- How about executing the following lines in `configure.py` instead of `Makefile_base` to improve the internal consistency further?
```
ifeq "$(filter -DGAMER_DEBUG, $(SIMU_OPTION))" "-DGAMER_DEBUG"
   NVCCFLAG_COM += -g -Xptxas -v   # "-G" may cause the GPU Poisson solver to fail
endif
```
- Modify `CUDAFLAG` in all `*.config` as follows.
  - Replace `--ccbin` with `-ccbin`
  - Combine it with `NVCCFLAG_COM` by
  ```
  NVCCFLAG_COM -O3
  #NVCCFLAG_COM --ccbin YOUR_HOST_COMPILER
  #NVCCFLAG_COM --ccbin CC
  ```
  - Remove `CUDAFLAG` from `configure.py` since it's now part of `NVCCFLAG_COM`.
- Remove the following lines from `*.config` as they are deprecated.
```
#NVCCFLAG_COM -Xcompiler
#NVCCFLAG_COM -Xopencc -OPT:Olimit=0
```

**Comment by ChunYen-Chen** [comment]
@hyschive Thanks for your review. I have changed the code according to the comments.

**Review by hyschive — APPROVED** [review]
@ChunYen-Chen Looks good. I'll merge it right away!

There is a typo `GPU_COMPUTABILITY` in `template.config` but I'll fix it by myself.

**Pull request merged**